ferrocene/tools/self-test/src/linkers.rs

```diff
@@ -1,71 +1,257 @@
+                        if std::env::var("FST_PRINT_DETAILED_ERRORS").is_ok() {
+                            reporter.note(&format!("`{compiler_name}` failed with {e}"));
+                        }
+                    }
+                }
             }
         }
+        return Err(Error::SuitableCCompilerNotFound { target: target.triple.to_owned() });
     }
 
     Ok(())
 }
 
-fn check_bundled_lld(
-    reporter: &dyn Reporter,
-    sysroot: &Path,
-    targets: &[String],
-) -> Result<PathBuf, Error> {
+/// Check if the given system C compiler works.
+///
+/// We are given a path to the real `ld.lld`, but we also test it with a fake
+/// `ld.lld` as well.
+///
+/// We are also given a path to a fresh temporary directory we can put source
+/// code in.
+///
+/// Returns the path to the C compiler, and a list of arguments that the C
+/// compiler gives to the linker.
+fn check_system_compiler(
+    environment: &Environment,
+    target: &str,
+    compiler_name: &str,
+    lld_dir: &Path,
+    temp_dir: &Path,
+) -> Result<(PathBuf, Vec<String>), Error> {
+    let cc_path = find_binary_in_path(environment, &compiler_name)
+        .map_err(|error| Error::CCompilerNotFound { name: compiler_name.to_owned(), error })?;
+
+    // Part 1. Check with the real ld.lld - can we make a binary?
+
+    compile_c_program(&cc_path, lld_dir, temp_dir)?;
+
+    // Part 2. Make a fake linker, and get GCC to try and use it. What arguments
+    // does it give our fake linker?
+
+    let args_file = make_fake_linker(temp_dir)?;
+
+    let linker_args = check_compiler_linker_args(target, &cc_path, temp_dir, &args_file)?;
+
+    Ok((cc_path, linker_args))
+}
+
+/// Check if the given system C compiler works.
+///
+/// We are given a path to the real `ld.lld`.
+///
+/// We are also given a path to a fresh temporary directory we can put source
+/// code in.
+///
+/// Returns the path to the C compiler.
+fn compile_c_program(cc_path: &Path, lld_dir: &Path, temp_dir: &Path) -> Result<(), Error> {
+    // We need some C source code,
+    let c_source = r#"int main(void) { return 0; }"#;
+
+    // We need a temp directory we can save the output file to
+    let source_file = temp_dir.join("input.c");
+    let object_file = temp_dir.join("output.bin");
+    std::fs::write(&source_file, c_source.as_bytes()).expect("saving source code to temp file");
+
+    // We need to call the C compiler, telling it to use ld.lld and telling it where to find ld.lld
+    let args: Vec<OsString> = vec![
+        "-fuse-ld=lld".into(),
+        "-B".into(),
+        lld_dir.as_os_str().to_owned(),
+        source_file.as_os_str().to_owned(),
+        "-o".into(),
+        object_file.as_os_str().to_owned(),
+    ];
+    let mut cc_child = Command::new(cc_path);
+    cc_child.args(&args);
+
+    let _output = run_command(&mut cc_child).map_err(|error| {
+        let cc_name = cc_path.file_name().unwrap().to_str().unwrap();
```

Comment:
There is `Error::NonUtf8Path` for this.

Reply:
Returning a non-UTF-8 error here would hide the fact the actual problem was that "`run_command` failed". I'll unwrap_or in a replacement string for the unlikely event the C compiler path is not UTF-8.